cmd/kubelet/app/options/options.go

```diff
@@ -554,6 +554,7 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
 	fs.Int32Var(&c.PodsPerCore, "pods-per-core", c.PodsPerCore, "Number of Pods per core that can run on this Kubelet. The total number of Pods on this Kubelet cannot exceed max-pods, so max-pods will be used if this calculation results in a larger number of Pods allowed on the Kubelet. A value of 0 disables this limit.")
 	fs.BoolVar(&c.ProtectKernelDefaults, "protect-kernel-defaults", c.ProtectKernelDefaults, "Default kubelet behaviour for kernel tuning. If set, kubelet errors if any of kernel tunables is different than kubelet defaults.")
 	fs.StringVar(&c.ReservedSystemCPUs, "reserved-cpus", c.ReservedSystemCPUs, "A comma-separated list of CPUs or CPU ranges that are reserved for system and kubernetes usage. This specific list will supersede cpu counts in --system-reserved and --kube-reserved.")
+	fs.StringVar(&c.StaticCPUs, "static-cpus", c.StaticCPUs, "A comma-separated list of CPUs or CPU ranges that will be used exclusive assignemnt. This list restricts the list of CPUs available for exclusive assignemnt. Containers with guaranteed QOS will be assigned with CPUs only from this list.")
```

Comment:
I am trying to understand this proposal better.

Today, if I use either `--system-reserved/kube-reserved` OR `reserved-cpus`, I am reducing my node allocatable capacity.  If used in concert with cpu manager policy of `static`, the specified or low-numbered cores are excluded from exclusive scheduling BUT for non G QoS pods.

If I use `static-cpus`, I am not reducing allocatable, but I am implicitly restricting the number of G versus non G QoS pods I can run on a node without any corresponding signal back to the scheduler.  Is this an accurate understanding of the proposal?  The net is a kubelet could report allocatable cpu=6, but if static-cpus=1-4, its really allocatable=6 where 4 can be used by G pods, and 2 are left for non G pods.  This seems confusing, but its possible I am reading this wrong.

If you could show a concrete configuration of isolcpu, reserved-cpus, and static-cpus used in concert it would help.

Reply:
Another way to phrase this, prior to this `static-cpu` proposal the list of CPU ranges that could be handed out for exclusive assignment dynamically adjusted with the pods that were scheduled to the node.  This makes that dynamism go away, and restricts where non G pods are able to execute and fit.

Reply:
Thanks for the review @derekwaynecarr 

> I am trying to understand this proposal better.
> 
> Today, if I use either `--system-reserved/kube-reserved` OR `reserved-cpus`, I am reducing my node allocatable capacity. If used in concert with cpu manager policy of `static`, the specified or low-numbered cores are excluded from exclusive scheduling BUT for non G QoS pods.
> 
This is true from k8s perspective, however, in case isolcpus kernel option is configured, it will prevent process running in these non G QoS pods to execute on the isolated cpus.
If there are total 6 cpus, and we will configure --reserved-cpus=0-1, isolcpus=2-5, non G pods will be able to execute only on 0-1

> If I use `static-cpus`, I am not reducing allocatable, but I am implicitly restricting the number of G versus non G QoS pods I can run on a node without any corresponding signal back to the scheduler. Is this an accurate understanding of the proposal? The net is a kubelet could report allocatable cpu=6, but if static-cpus=1-4, its really allocatable=6 where 4 can be used by G pods, and 2 are left for non G pods. This seems confusing, but its possible I am reading this wrong.
Yes, that's correct, however, static-cpus only applies to G QoS pods, the same way as reserved-cpus only applies to G QoS pods. 

> 
> If you could show a concrete configuration of isolcpu, reserved-cpus, and static-cpus used in concert it would help.

I think there will be no need to configure `reserved-cpus` with the new `static-cpus` option. 
If we have cpus=6, (kernel) isolcpus=2-5, static-cpus=2-5 

You are right about the scheduler. I don't know how to solve this. I didn't find a way to account for only G CPUs with the `static-cpus` option. However, I thought that this will mimic the behavior of the Affinity Manager, which will error if it couldn't align pod CPUs to the correct numa node.

Reply:
> Another way to phrase this, prior to this `static-cpu` proposal the list of CPU ranges that could be handed out for exclusive assignment dynamically adjusted with the pods that were scheduled to the node. This makes that dynamism go away, and restricts where non G pods are able to execute and fit.

I think the non G pods are not affected by this change, from k8s perspective. At least I didn't intend to change this behavior

As for the dynamism, you're right, but it is also gone when kernel `isolcpus` option is used in a combination with `reserved-cpus`.

My take on this, and it may be wrong, is that those who configure `isolcpus` on the node know that they are eventually limiting their G pods to only this list of CPUs. However, without the `static-cpus` option they didn't have a way to effectively restrict the pool of the G CPUs to only `isolcpus` without reducing the nodes allocatable capacity.

Reply:
one possible problem with this approach: can the static cpu set ever be used by the non G pod? say when there is G pods, are the specified static cpu set be used by the non G pods?

Reply:
another possible issue: say you have 6 allocable cpu on the system but static-cpu=4-5; a pod which request 4 static cpu can be schedule to this node; but the 4>2 and kubelet will fail to accommodate this pod. Right?